src/ess_rhat.jl

```diff
@@ -239,19 +257,184 @@ The ESS and ``\\widehat{R}`` values can be computed for the following estimators
+        throw(ArgumentError("the estimator $estimator is not yet supported by `ess`"))
     end
-    values = ess_rhat(Statistics.mean, x; kwargs...)
-    return values
+    return _ess(Val(:basic), x; kwargs...)
+end
+function _ess(kind::Val, samples::AbstractArray{<:Union{Missing,Real},3}; kwargs...)
+    return first(_ess_rhat(kind, samples; kwargs...))
+end
+function _ess(
+    ::Val{:tail},
+    x::AbstractArray{<:Union{Missing,Real},3};
+    tail_prob::Real=1//10,
+    kwargs...,
+)
+    # workaround for https://github.com/JuliaStats/Statistics.jl/issues/136
+    T = Base.promote_eltype(x, tail_prob)
+    pl = convert(T, tail_prob / 2)
+    pu = convert(T, 1 - tail_prob / 2)
+    S_lower = _ess(Base.Fix2(Statistics.quantile, pl), x; kwargs...)
+    S_upper = _ess(Base.Fix2(Statistics.quantile, pu), x; kwargs...)
+    return map(min, S_lower, S_upper)
 end
+
+"""
+    rhat(samples::AbstractArray{Union{Real,Missing},3}; kind::Symbol=:rank, split_chains=2)
+
+Compute the ``\\widehat{R}`` diagnostics for each parameter in `samples` of shape
+`(chains, draws, parameters)`.[^VehtariGelman2021]
+
+`kind` indicates the kind of ``\\widehat{R}`` to compute (see below).
+
+$_DOC_SPLIT_CHAINS
+
+See also [`ess`](@ref), [`ess_rhat`](@ref), [`rstar`](@ref)
+
+## Kinds of ``\\widehat{R}``
+
+The following `kind`s are supported:
+- `:rank`: maximum of ``\\widehat{R}`` with `kind=:bulk` and `kind=:tail`.
+- `:bulk`: basic ``\\widehat{R}`` computed on rank-normalized draws. This kind diagnoses
+    poor convergence in the bulk of the distribution due to trends or different locations of
+    the chains.
+- `:tail`: ``\\widehat{R}`` computed on draws folded around the median and then
+    rank-normalized. This kind diagnoses poor convergence in the tails of the distribution
+    due to different scales of the chains.
+- `:basic`: Classic ``\\widehat{R}``.
+
+[^VehtariGelman2021]: Vehtari, A., Gelman, A., Simpson, D., Carpenter, B., & Bürkner, P. C. (2021).
+    Rank-normalization, folding, and localization: An improved ``\\widehat {R}`` for
+    assessing convergence of MCMC. Bayesian Analysis.
+    doi: [10.1214/20-BA1221](https://doi.org/10.1214/20-BA1221)
+    arXiv: [1903.08008](https://arxiv.org/abs/1903.08008)
+"""
+function rhat(
+    samples::AbstractArray{<:Union{Missing,Real},3}; kind::Symbol=:rank, kwargs...
+)
+    if kind === :rank
+        return _rhat(Val(:rank), samples; kwargs...)
+    elseif kind === :bulk
+        return _rhat(Val(:bulk), samples; kwargs...)
+    elseif kind === :tail
+        return _rhat(Val(:tail), samples; kwargs...)
+    elseif kind === :basic
+        return _rhat(Val(:basic), samples; kwargs...)
+    else
+        return throw(ArgumentError("the `kind` `$kind` is not supported by `rhat`"))
+    end
+end
+function _rhat(
+    ::Val{:basic}, chains::AbstractArray{<:Union{Missing,Real},3}; split_chains::Int=2
+)
+    # compute size of matrices (each chain may be split!)
+    niter = size(chains, 1) ÷ split_chains
```

Comment:
It's a bit annoying that the code has (?) to be duplicated. Hopefully at least the tests make it unlikely that they diverge in a noticeable way.

Reply:
Yeah, I think we may need a refactor of the code that computes the various variances when we tackle #23, and hopefully then we can eliminate at least part of the redundancy. But yes, the tests should prevent divergences.